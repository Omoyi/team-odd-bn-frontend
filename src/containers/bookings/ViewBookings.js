import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Dashboard from '../../views/Dashboard/sidebar';
import { getAllBookings } from '../../redux/actions/bookings/getAllBookingsActions';
import { Paginate } from '../../helpers/Paginate';
import { changePageNo } from '../../redux/actions/PaginationAction';
import '../../assets/css/bookings/userBookings.scss';
import icon1 from '../../assets/images/ViewAllBookings/icon1.png';
import icon2 from '../../assets/images/ViewAllBookings/icon2.png';
import icon3 from '../../assets/images/ViewAllBookings/icon3.png';
import icon4 from '../../assets/images/ViewAllBookings/icon4.png';
import { likeAction, dislikeAction } from '../../redux/actions/likeAndDislikeAction';
import { viewActionAccommodation } from '../../redux/actions/allAccommodation';
import likeIcon from '../../assets/images/like_icon/thumbs-up_32.png';
import disLikeIcon from '../../assets/images/dislike_icon/thumbs-down_32.png';
import joinLikesAccomodation from '../../helpers/joinLikesAccomodation'
import { getCitiesAction } from '../../redux/actions/createAccommodationActions';
import { openSingleBooking, getSingleBooking } from '../../redux/actions/bookings/viewSingleBookingActions';
import ViewSingleBooking from '../../containers/bookings/viewSingleBooking';
import { onEditInputAction } from '../../redux/actions/ratingAction'
import popUpAction from '../../redux/actions/popUpAction';
import PopUp from '../../components/PopUp';
import RatesBookings from '../../views/bookings/RatesBookings';
import Pagination from 'custom_react_pages'

export class ViewBookings extends Component {

  state = {
    modal: false,
    index: null,
  }
  async componentDidMount() {    
    this.props.getCitiesAction();
    const { viewActionAccommodation } = this.props;
    viewActionAccommodation();
    await this.props.getAllBookings();
    const { allAccomodation } = this.props.accommodation;
    this.setState({ allAccomodation });
  }
  
  handleSubmit  = (e, id, index) => {  
    e.preventDefault();
    this.setState(prevState=>(
      {
        modal: !prevState.modal,
        index
      }
    ))
    this.props.openSingleBooking(true);
    this.props.getSingleBooking(id.id);
  }

  handleClick = async(like, item) =>{       
    const { likeAction,viewActionAccommodation, dislikeAction } = this.props;

    if(like === "true"){
    const passData =  await likeAction({like: true, id: item})
      const updateView = await viewActionAccommodation()
      return  { passData, updateView };
    }
    else if(like === "false"){
      const passData = await dislikeAction( {dislike: true, id: item})
      const updateView =  await viewActionAccommodation();
      return  { passData, updateView };
    }
    return await viewActionAccommodation();
  }

  render() {
  
    const {
      allBookings, pageNo, itemsPerPage, changePageNo, allAccomodation, popUpAction, display, onEditInputAction
    } = this.props;
    const { accommodations, allLikes } = allAccomodation;

    allLikes.map((like, index) => Object.assign(accommodations[index], like));
    
    const accommodationsData = allBookings ? allBookings.data.myBookingsresult.map(item=>item):[];
    
    joinLikesAccomodation(accommodations,allLikes).map((item)=>{
      accommodationsData.forEach(element => {
        if(item.id === element.room.accommodationId){
          Object.assign(element,item)
        }
      });
    })
  
    const chunkBookings = Paginate(accommodationsData, itemsPerPage);
    const bookingsPages = { ...chunkBookings };
    const currentPage = bookingsPages[pageNo];
    

    return (
      <Dashboard>
        <div className="userBookingContainer">
          <div className="userBookingsHeader">
            <div className="bookingTitle">Bookings</div>
          </div>
          {this.props.allBookings ? (
            <div className="userBookingsContainer">
              {this.state.modal && (
                <ViewSingleBooking
                  handleUpdateModal={item => this.setState({ modal: !item })}
                  myIndex={this.state.index}
                  modal={this.state.modal}
                />
              )}

              {accommodationsData.length > 0 && (
                <Pagination
                  itemsPerPage={6}
                  next="next"
                  prev="previous"
                  data={accommodationsData.length ? accommodationsData : []}
                  oneItem={(booking, index) => (
                    <div key={index} className="singleBooking">
                      <div
                        className="bookingImage"
                        onClick={e => {
                          this.handleSubmit(e, booking, index)
                        }}
                      >
                        <img
                          src={
                            booking.room.accommodation.imagesAccommodation[0]
                              .imageUrl
                          }
                        />
                      </div>
                      <div className="bookingDetails">
                        <div className="bookingHeader">
                          {booking.room.accommodation.name}
                        </div>
                        <div className="checkInBookings">
                          <img src={icon2} alt="icon1" />
                          <p>
                            Check-in date:
                            {booking.checkInDate.slice(0, 10)}
                          </p>
                        </div>
                        <div className="checkOutBookings">
                          <img src={icon1} alt="icon2" />
                          <p>
                            Check-out date:
                            {booking.checkOutDate.slice(0, 10)}
                          </p>
                        </div>
                        <div className="roomBookings">
                          <img src={icon3} alt="icon3" />
                          <p>
                            {`${booking.room.roomType} | ${booking.room.name}`}
                          </p>
                        </div>
                        <div className="cityBooking">
                          <img src={icon4} alt="icon4" />
                          <p>{booking.room.accommodation.city.city}</p>
                        </div>

                        <div className="booking-like-contents">
                          <section className="common-content like-dislikes">
                            <div className="like-dislike-contents like-container">
                              <img
                                onClick={() => {
                                  const item = booking.room.accommodationId

                                  return this.handleClick('true', item)
                                }}
                                className="likes_icon_img"
                                src={likeIcon}
                                alt="like icon"
                              />
                            </div>
                            <div className="like-dislike-contents dislike-container">
                              <img
                                onClick={() => {
                                  const item = booking.room.accommodationId
                                  this.handleClick('false', item)
                                }}
                                className="dislike_icon"
                                src={disLikeIcon}
                                alt="dislike icon"
                              />
                            </div>
                          </section>
                          <section className="common-content likes_dislikes_data">
                            <span className="likes_side">
                              {booking.likeCounter}
                            </span>
                            <span className="dislikes_side">
                              {booking.dislikeCounter}
                            </span>
                          </section>

                          <div className="rateBookingsBtn">
                            <button
                              className="rateBookingBtn"
                              type="button"
                              className="form-btn-open"
                              onClick={() => {
                                popUpAction({
                                  currentPopUp: 'rate',
                                  rate: 'flex',
                                })
                                onEditInputAction({
                                  accommodationId: booking.id,
                                })
                              }}
                            >
                              Rate
                            </button>
                          </div>
                        </div>

                        <div className="createdOnBooking">
                          <p>
                            Request Date:
                            {booking.createdAt.slice(0, 10)}
                          </p>
                        </div>
                      </div>
                    </div>
                  )}
                />
              )}
            </div>
          ) : (
            <div className="empty-bookings-container">
              <p className="error-message-booking">
                You do not have bookings click
                <Link to="/accommodations"> here </Link>
                to book new one
              </p>
            </div>
          )}

          <PopUp popUp={<RatesBookings />} display={display.rate} />
        </div>
      </Dashboard>
    )
  }
}
export const mapStateToProps = (state, { userLikesAndDislike }) => ({
  allBookings: state.bookings.bookings.allBookings,
  pageNo: state.pagination.pageNo,
  itemsPerPage: state.pagination.itemsPerPage,
  likeDislikeState: userLikesAndDislike,
  allAccomodation: state.allAccomodation,
  bookings: state.bookings,
  display: state.popUpsDisplay,
});

export default connect(mapStateToProps, {
  getAllBookings, changePageNo, viewActionAccommodation, likeAction, 
  dislikeAction, openSingleBooking, getSingleBooking, getCitiesAction, 
  popUpAction, onEditInputAction
})(
  ViewBookings,
);
