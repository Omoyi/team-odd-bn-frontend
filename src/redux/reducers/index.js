import { combineReducers } from 'redux';
import socialAuth from './socialAuthReducer';
import authReducers from './authReducers';
import profileReducers from './ProfileReducers';
import updateProfileReducer from './updateProfileReducer';
import getAllUsersReducer from './getAllUsersReducers';
import signoutReducer from './signoutReducer';
import tripReducers from './tripReducers';
import getAllCitiesReducers from './getAllCitiesReducers';
import accommodationReducer from './accommodation';
import multicityReducer from './multicityReducer';
import adminReducer from './admin/adminReducer';
import viewAccomodationReducer from './accommodation/viewAccomodationReducer';
import PaginationReducer from './PaginationReducer';
import searchReducer from './searchReducers/searchReducers';

import bookingReducer from './bookings';

export default combineReducers({
  viewProfile: profileReducers,
  updateProfile: updateProfileReducer,
  allUsers: getAllUsersReducer,
  socialAuth,
  auth: authReducers,
  signout: signoutReducer,
  trips: tripReducers,
  allCities: getAllCitiesReducers,
  accommodation: accommodationReducer,
  multicityTrip: multicityReducer,
  admin: adminReducer,
  allAccomodation: viewAccomodationReducer,
  pagination: PaginationReducer,
  bookings: bookingReducer,
  search: searchReducer,
});
