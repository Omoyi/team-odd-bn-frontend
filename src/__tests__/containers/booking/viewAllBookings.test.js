import React from 'react';
import { shallow, mount } from 'enzyme';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import {
  tempViewAllBooking,
  initiaState,
  viewBookingsMockData,
} from '../../../__mocks__/booking/booking';
import rootReducer from '../../../redux/reducers/index';
import { ViewBookings } from '../../../containers/bookings/ViewBookings.js';


const middlewares = [thunk];
const testStore = (state) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, state);
};


describe('Expect to view bookings', () => {
  const wrapper = shallow(<ViewBookings {...viewBookingsMockData} />);
  it('should view all bookings successfully', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
