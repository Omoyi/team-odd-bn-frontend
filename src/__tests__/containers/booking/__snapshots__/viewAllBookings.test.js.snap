// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect to view bookings should view all bookings successfully 1`] = `ShallowWrapper {}`;
